crates/red_knot/src/salsa_db/semantic/module.rs

```diff
@@ -0,0 +1,886 @@
+use std::fmt::Formatter;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use crate::module::ModuleName;
+use crate::salsa_db::source::File;
+use crate::symbols::Dependency;
+
+use super::Db;
+use super::Jar;
+
+/// ID uniquely identifying a module.
+#[salsa::interned(jar=Jar)]
+pub struct Module {
+    #[return_ref]
+    name: ModuleName,
+}
+
+#[salsa::tracked(jar=Jar)]
+pub struct ResolvedModule {
```

Comment:
@AlexWaygood this is where I'm currently landing on a Salsa design for a module resolver. I think it would simplify a lot for you because you no longer need to think about invalidation, Salsa will take care of you. The only thing necessary for this to work is that you use `db.file(path).exists()` to test if a file exists.

Reply:
Ah, thanks for the ping. Yes, this indeed does make the code look a _lot_ cleaner! It was making my head hurt a little bit to see all the cache-checking stuff right alongside the search-path semantics in `resolve_module()`